ramips: Add support for D-Link DWR-118-A2

CHKDSK88 opened the pull request:
The DWR-118-A2 Wireless Router is based on the MT7620A SoC.

Specification:

- MediaTek MT7620A (580 Mhz)
- 128 MB of RAM
- 16 MB of FLASH
- 1x 802.11bgn radio
- 1x 802.11ac radio (MT7612EN)
- 4x 10/100 Mbps Ethernet (1 WAN and 3 LAN)
- 1x 10/100/1000 Mbps Ethernet (1 LAN)
- 2x external, non-detachable antennas
- 1x USB 2.0
- UART (J1) header on PCB (57600 8n1)
- 7x LED (5x GPIO-controlled), 2x button
- JBOOT bootloader

Known issues:
- GELAN not working
- flash is very slow

The status led has been assigned to the dwr-118-a2:green:internet led.
At the end of the boot it is switched off and is available for other
operation. Work correctly also during sysupgrade operation.

Installation:
Apply factory image via http web-gui or JBOOT recovery page

How to revert to OEM firmware:
- push the reset button and turn on the power. Wait until LED start
  blinking (~10sec.)
- upload original factory image via JBOOT http (IP: 192.168.123.254)

Signed-off-by: Cezary Jackiewicz <cezary@eko.one.pl>
Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

obsy committed:
ramips: Add Marvell PHY

Added support for Marvell PHY used on some devices.

Signed-off-by: Cezary Jackiewicz <cezary@eko.one.pl>

obsy committed:
ramips: Add support for D-Link DWR-118-A2

The DWR-118-A2 Wireless Router is based on the MT7620A SoC.

Specification:

- MediaTek MT7620A (580 Mhz)
- 128 MB of RAM
- 16 MB of FLASH
- 1x 802.11bgn radio
- 1x 802.11ac radio (MT7612EN)
- 4x 10/100 Mbps Ethernet (1 WAN and 3 LAN)
- 1x 10/100/1000 Mbps Ethernet (1 LAN)
- 2x external, non-detachable antennas
- 1x USB 2.0
- UART (J1) header on PCB (57600 8n1)
- 7x LED (5x GPIO-controlled), 2x button
- JBOOT bootloader

Known issues:
- GELAN not working
- flash is very slow

The status led has been assigned to the dwr-118-a2:green:internet led.
At the end of the boot it is switched off and is available for other
operation. Work correctly also during sysupgrade operation.

Installation:
Apply factory image via http web-gui or JBOOT recovery page

How to revert to OEM firmware:
- push the reset button and turn on the power. Wait until LED start
  blinking (~10sec.)
- upload original factory image via JBOOT http (IP: 192.168.123.254)

Signed-off-by: Cezary Jackiewicz <cezary@eko.one.pl>
Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

mkresin commented on target/linux/ramips/files-4.14/drivers/net/ethernet/mediatek/gsw_mt7620.c:
```diff
@@ -157,6 +157,11 @@ static void mt7620_hw_init(struct mt7620_gsw *gsw, struct device_node *np)
 	mtk_switch_w32(gsw, 0x7f7f7fe0, 0x0010);
 
+	/* enable port 0 */
+	if (of_property_read_bool(np, "mediatek,enableport0")) {
```
Considering that we added the same for port4 the last time and now for port0, wouldn't it make more sense to introduce a `mediatek,portdisable` property?

The idea is to define a binary mask of ports which should be disabled. If the port has a 0 value it should be enabled.

We have such code already in `target/linux/ramips/files-4.14/drivers/net/ethernet/mediatek/esw_rt3050.c` (12758e32d3309c5147bd342525320ad704e74a07).

CHKDSK88 committed:
ramips: mt7620: enable all ports unconditionally

This patch make all mt7620 ephy ports turned on.
It is necessary for some JBOOT devices.

Signed-off-by: Pawel Dembicki <paweldembicki@gmail.com>

CHKDSK88 commented:
@mkresin 

I agree with You and I reworked my PR.

mkresin commented:
Hey Pawel,

your port_disable/enable code doesn't really work and is a little bit overenginered.

First of all, it doesn't make much sense to store the port_disable map in the mt7620_gsw struct if it is used only one time in the code. It would make more sense to read, use and drop the port_disable map instead.

The `1 << i`can be simplified by using `BIT(i)` instead. At least for me it is way better to read this way.

If you are doing a binary AND, you have to use a single `&`. What your code really does is `if (port_disable == true AND (1 << i) == true)`. Means, it worked only by accident for you.

I don't like the idea of setting the whole register value while we are only interested in the power up/down bit.

I played a bit with your code with the following result:

```diff
diff --git a/target/linux/ramips/files-4.14/drivers/net/ethernet/mediatek/gsw_mt7620.c b/target/linux/ramips/files-4.14/drivers/net/ethernet/mediatek/gsw_mt7620.c
index a4602c4..fedb4bf 100644
--- a/target/linux/ramips/files-4.14/drivers/net/ethernet/mediatek/gsw_mt7620.c
+++ b/target/linux/ramips/files-4.14/drivers/net/ethernet/mediatek/gsw_mt7620.c
@@ -63,6 +63,7 @@ static irqreturn_t gsw_interrupt_mt7620(int irq, void *_priv)
 static void mt7620_hw_init(struct mt7620_gsw *gsw, struct device_node *np)
 {
 	u32 is_BGA = (rt_sysc_r32(0x0c) >> 16) & 1;
+	u32 port_disable;
 
 	rt_sysc_w32(rt_sysc_r32(SYSC_REG_CFG1) | BIT(8), SYSC_REG_CFG1);
 	mtk_switch_w32(gsw, mtk_switch_r32(gsw, GSW_REG_CKGCR) & ~(0x3 << 4), GSW_REG_CKGCR);
@@ -166,9 +167,24 @@ static void mt7620_hw_init(struct mt7620_gsw *gsw, struct device_node *np)
 		_mt7620_mii_write(gsw, 4, 30, 0xa000);
 		_mt7620_mii_write(gsw, 4, 4, 0x05e1);
 		_mt7620_mii_write(gsw, 4, 16, 0x1313);
-		_mt7620_mii_write(gsw, 4, 0, 0x3100);
 		pr_info("gsw: setting port4 to ephy mode\n");
 	}
+
+	if (!of_property_read_u32(np, "mediatek,portdisable", &port_disable)) {
+		u32 port;
+		u32 val;
+
+		for (port = 0; port < 5; port++) {
+			val = _mt7620_mii_read(gsw, port, 0);
+
+			if (port_disable & BIT(port))
+				val |= BIT(11);
+			else
+				val &= ~BIT(11);
+
+			_mt7620_mii_write(gsw, port, 0, val);
+		}
+	}
 }
 
 static const struct of_device_id mediatek_gsw_match[] = {
```

I was somehow surpised that we're only facing the disabled ports issue with mt7620. After checking the mt7621 gsw driver, I'm no longer surprised.

During hw_init all phys are [turned off](https://github.com/openwrt/openwrt/blob/fdd6c556ab42d040efcc92a53149e3f8ec357f69/target/linux/ramips/files-4.14/drivers/net/ethernet/mediatek/gsw_mt7621.c#L90) and finally all phys are [turned on](https://github.com/openwrt/openwrt/blob/fdd6c556ab42d040efcc92a53149e3f8ec357f69/target/linux/ramips/files-4.14/drivers/net/ethernet/mediatek/gsw_mt7621.c#L189), similar to what I suggested to do via mail some time ago.

I would suggest to copy/paste the mt7621 gsw "turn on all PHYs" code instead and enable all phys unconditionally. It keeps the drivers at least a bit in sync and should be the best option for now.

We can add a clean fix as soon as we use the [distributed switch architecture device tree binding](https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/net/dsa/dsa.txt#L83), where each port is expressed as own device tree node. With that syntax we can enable all listed ports and disable all not listed ports.

CHKDSK88 commented:
@mkresin 

I'am sorry for &/&& mistake. 
Solution from mt7621 is the best method at this moment. I reworked PR again.



Maybe after reworking driver to DSA will be possible to enable gigabit port in DWR-118. At this moment is impossible to set address = 0 to external phy (phy of gmac port 4 have address zero and now is masked by ephy). I try to change this in my another branch but i stuck at dead end.

mkresin commented:
>  I reworked PR again.

Are you going to rework the PR or have you forgot to pus you latest changes to the PR?

xabolcs commented:
@CHKDSK88 , how does PR #1247 relate to "GELAN not working"?


@mkresin , it seems to updated to me: the middle commit is simplified to "switch all port off and on". Are you missing something?

danielg4 commented:
@xabolcs #1247 does not relate to this, and from what I can tell, neither does #1238

mkresin commented:
@CHKDSK88 I dropped the "turn off all PHYs" code. It isn't strictly required to fix your issue, isn't mentioned in the commit message and I don't like to risk possible new breakage due to disabling all phys. Disabling all PHYs might work but better safe than sorry.

I applied some changes to dts which are required due to my recent ramips cleanup commits.

As usual, you find the results in my staging tree. Please test an image build from my staging tree and let me know the result. I'll push the commits afterwards.

CHKDSK88 commented:
Thank You. I check after FS#1754 fix.

CHKDSK88 commented:
@mkresin 

I tested. Everything work ok.
